src/content/en/fundamentals/media/fast-playback-with-video-preload.md

```diff
@@ -0,0 +1,578 @@
+
+    // Fetch the rest of the video when user starts playing video.
+    video.addEventListener('playing', fetchRemainder, { once: true });
+  }
+
+  function fetchRemainder() {
+    fetch('file.webm', { headers: { range: 'bytes=567140-1196488' } })
+    .then(response => response.arrayBuffer())
+    .then(data => {
+      const sourceBuffer = mediaSource.sourceBuffers[0];
+      sourceBuffer.appendBuffer(data);
+    });
+  }
+</script>
+```
+
+### Battery awareness
+
+I would suggest you take into account the battery level of user's device before
+thinking about preloading a video in order to preserve battery life when the
+power level is low.
+
+Please disable preload or at least preload a lower resolution video when the
+device is running out of battery.
+
+
+```
+if ('getBattery' in navigator) {
+  navigator.getBattery()
+  .then(battery => {
+    // If battery is charging or battery level is high enough
+    if (battery.charging || battery.level > 0.15) {
+      // TODO: Preload the first segment of a video.
+    } else {
+      // TODO: Enable some kind of Power Saving mode.
+    }
+  });
+} else {
+  // Battery Status API is not supported.
+  // TODO: Preload the first segment of a video.
+}
+```
+
+
+### Pre-cache multiple first segments
+
+Now, what if I want to speculatively pre-load some media content without
+knowing which piece of media the user will eventually pick. If user is on a web
+page that contains 10 videos, we probably have enough memory to fetch one
+segment file from each but we should definitely not create 10 hidden video
+elements and 10 `MediaSource` objects and start feeding that data.
+
+The 2-parts example below shows you how to pre-cache multiple first segments of
+video using the powerful and easy-to-use Cache API. Note that something similar
+can be achieved with IndexedDB as well. We're not using service workers yet as
+the Cache API is also accessible from the Window object.
+
+#### Fetch and cache
+
+```
+const videoFileUrls = [
+  'bat_video_file_1.webm',
+  'cow_video_file_1.webm',
+  'dog_video_file_1.webm',
+  'fox_video_file_1.webm',
+];
+
+// Let's create a video pre-cache and store all first segments of videos inside.
+window.caches.open('video-pre-cache')
+.then(cache => Promise.all(videoFileUrls.map(videoFileUrl => fetchAndCache(videoFileUrl, cache))));
+
+function fetchAndCache(videoFileUrl, cache) {
+  // Check first if video is in the cache.
+  return cache.match(videoFileUrl)
+  .then(cacheResponse => {
+    // Let's return cached response if video is already in the cache.
+    if (cacheResponse) {
+      return cacheResponse;
+    }
+    // Otherwise, fetch the video from the network.
+    return fetch(videoFileUrl)
+    .then(networkResponse => {
+      // Add the response to the cache and return network response in parallel.
+      cache.put(videoFileUrl, networkResponse.clone());
+      return networkResponse;
+    });
+  });
+}
```

Comment:
I wonder if these examples would be a lot easier to read as async functions. That presumes developers understand them though.